File: AnyTorah/API/DedicationService.swift
import Foundation

struct Dedication: Identifiable {
    var id: String { "\(date.timeIntervalSince1970)-\(period)-\(dedicatedBy)" }
    let date: Date
    let endDate: Date         // inclusive; equals `date` for a single-day dedication
    let dedicatedBy: String
    let honoreeName: String
    let period: String        // "today" | "week" | "month" — display wording only, see below
    let preposition: String
    let occasion: String
    let displayText: String?
    let photoURL: String?

    var periodTitle: String {
        switch period {
        case "week":  return "This Week's Learning"
        case "month": return "This Month's Learning"
        default:      return "Today's Learning"
        }
    }

    var formattedMessage: String {
        if let override = displayText, !override.isEmpty { return override }
        let periodPhrase: String
        switch period {
        case "week":  periodPhrase = "This week's learning"
        case "month": periodPhrase = "This month's learning"
        default:      periodPhrase = "Today's learning"
        }
        var parts = ["\(periodPhrase) with AnyTorah is dedicated by \(dedicatedBy)"]
        if !preposition.isEmpty  { parts.append(preposition) }
        if !honoreeName.isEmpty  { parts.append(honoreeName) }
        if !occasion.isEmpty     { parts.append(occasion) }
        return parts.joined(separator: " ") + "."
    }
}

enum DedicationService {
    private static let supabaseURL = "https://zewdazoijdpakugfvnzt.supabase.co"
    static let anonKey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inpld2Rhem9pamRwYWt1Z2Z2bnp0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzQ0NzIwODYsImV4cCI6MjA5MDA0ODA4Nn0.HJxIG18vEpt-exzoQwRLeXiKLAinWfBl7gMORKjxIz8"

    // Fetches the most relevant active dedication. "Active" means today falls in
    // [date, endDate], checked server-side by the query below. When more than one
    // dedication is active at once (an admin-acknowledged overlap — see
    // dedication-form.html's approval-time warning), picks by `period`'s display tier
    // (today > week > month), then most-recently-created.
    static func fetch() async -> Dedication? {
        let todayStr = todayDateString()

        let urlStr = "\(supabaseURL)/rest/v1/dedications"
            + "?date=lte.\(todayStr)"
            + "&end_date=gte.\(todayStr)"
            + "&status=eq.approved"
            + "&for_anytorah=eq.true"
            + "&select=date,end_date,dedicated_by,honoree_name,period,preposition,occasion,display_text,photo_url"
            + "&order=date.desc,id.desc"
            + "&limit=10"
        guard let url = URL(string: urlStr) else { return nil }
        var request = URLRequest(url: url)
        request.setValue(anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(anonKey)", forHTTPHeaderField: "Authorization")

        guard let (data, response) = try? await URLSession.shared.data(for: request),
              (response as? HTTPURLResponse)?.statusCode == 200,
              let rows = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]]
        else { return nil }

        return rows
            .compactMap { decode($0) }
            .sorted { periodPriority($0.period) > periodPriority($1.period) }
            .first
    }

    private static func decode(_ row: [String: Any]) -> Dedication? {
        guard let dateStr = row["date"] as? String,
              let date = parseDate(dateStr),
              let dedicatedBy = row["dedicated_by"] as? String,
              !dedicatedBy.isEmpty
        else { return nil }
        let endDate = (row["end_date"] as? String).flatMap(parseDate) ?? date
        return Dedication(
            date: date,
            endDate: endDate,
            dedicatedBy: dedicatedBy,
            honoreeName: (row["honoree_name"] as? String) ?? "",
            period:      (row["period"] as? String) ?? "today",
            preposition: (row["preposition"] as? String) ?? "",
            occasion:    (row["occasion"] as? String) ?? "",
            displayText: (row["display_text"] as? String).flatMap { $0.isEmpty ? nil : $0 },
            photoURL:    (row["photo_url"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        )
    }

    private static func periodPriority(_ period: String) -> Int {
        switch period {
        case "today": return 3
        case "week":  return 2
        default:      return 1
        }
    }

    private static func parseDate(_ str: String) -> Date? {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f.date(from: str)
    }

    static func formatDate(_ date: Date) -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f.string(from: date)
    }

    static func todayDateString() -> String { formatDate(Date()) }
}

File: AnyTorahAndroid/app/src/main/java/com/anytorah/api/DedicationService.kt
package com.anytorah.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.LocalDate

data class Dedication(
    val date: LocalDate,
    val endDate: LocalDate,   // inclusive; equals `date` for a single-day dedication
    val dedicatedBy: String,
    val honoreeName: String,
    val period: String,        // "today" | "week" | "month" — display wording only, see below
    val preposition: String,
    val occasion: String,
    val displayText: String?,
    val photoURL: String?
) {
    val periodTitle: String get() = when (period) {
        "week"  -> "This Week's Learning"
        "month" -> "This Month's Learning"
        else    -> "Today's Learning"
    }

    val formattedMessage: String get() {
        displayText?.takeIf { it.isNotBlank() }?.let { return it }
        val periodPhrase = when (period) {
            "week"  -> "This week's learning"
            "month" -> "This month's learning"
            else    -> "Today's learning"
        }
        val parts = mutableListOf("$periodPhrase with AnyTorah is dedicated by $dedicatedBy")
        if (preposition.isNotBlank()) parts.add(preposition)
        if (honoreeName.isNotBlank()) parts.add(honoreeName)
        if (occasion.isNotBlank())    parts.add(occasion)
        return parts.joinToString(" ") + "."
    }
}

object DedicationService {
    private const val SUPABASE_URL = "https://zewdazoijdpakugfvnzt.supabase.co"
    const val ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Inpld2Rhem9pamRwYWt1Z2Z2bnp0Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzQ0NzIwODYsImV4cCI6MjA5MDA0ODA4Nn0.HJxIG18vEpt-exzoQwRLeXiKLAinWfBl7gMORKjxIz8"

    // Fetches the most relevant active dedication. "Active" means today falls in
    // [date, endDate], checked server-side by the query below. When more than one
    // dedication is active at once (an admin-acknowledged overlap — see
    // dedication-form.html's approval-time warning), picks by `period`'s display tier
    // (today > week > month), then most-recently-created.
    suspend fun fetch(): Dedication? = withContext(Dispatchers.IO) {
        try {
            val today = LocalDate.now()
            val urlStr = "$SUPABASE_URL/rest/v1/dedications" +
                "?date=lte.$today" +
                "&end_date=gte.$today" +
                "&status=eq.approved" +
                "&for_anytorah=eq.true" +
                "&select=date,end_date,dedicated_by,honoree_name,period,preposition,occasion,display_text,photo_url" +
                "&order=date.desc,id.desc" +
                "&limit=10"
            val connection = URL(urlStr).openConnection() as HttpURLConnection
            connection.setRequestProperty("apikey", ANON_KEY)
            connection.setRequestProperty("Authorization", "Bearer $ANON_KEY")
            connection.connectTimeout = 5000
            connection.readTimeout = 5000
            if (connection.responseCode != 200) return@withContext null
            val body = connection.inputStream.bufferedReader().readText()
            JSONArray(body)
                .let { arr -> (0 until arr.length()).mapNotNull { decode(arr.getJSONObject(it)) } }
                .sortedByDescending { periodPriority(it.period) }
                .firstOrNull()
        } catch (_: Exception) { null }
    }

    private fun decode(row: JSONObject): Dedication? {
        val dateStr = row.optString("date").takeIf { it.isNotBlank() } ?: return null
        val dedicatedBy = row.optString("dedicated_by").takeIf { it.isNotBlank() } ?: return null
        val date = LocalDate.parse(dateStr)
        val endDate = row.optString("end_date").takeIf { it.isNotBlank() }?.let { LocalDate.parse(it) } ?: date
        return Dedication(
            date        = date,
            endDate     = endDate,
            dedicatedBy = dedicatedBy,
            honoreeName = row.optString("honoree_name"),
            period      = row.optString("period").ifBlank { "today" },
            preposition = row.optString("preposition"),
            occasion    = row.optString("occasion"),
            displayText = row.optString("display_text").takeIf { it.isNotBlank() },
            photoURL    = row.optString("photo_url").takeIf { it.isNotBlank() }
        )
    }

    private fun periodPriority(period: String): Int = when (period) {
        "today" -> 3
        "week"  -> 2
        else    -> 1
    }
}
